File: src/synapsepm/network/SynLibInterface.php
<?php
declare(strict_types=1);

namespace synapsepm\network;

use pocketmine\network\mcpe\protocol\DataPacket;
use pocketmine\network\SourceInterface;
use pocketmine\Player;
use synapsepm\network\protocol\spp\RedirectPacket;
use synapsepm\Synapse;


class SynLibInterface implements SourceInterface {
    private $synapseInterface;
    private $synapse;

    public function __construct(Synapse $synapse, SynapseInterface $interface) {
        $this->synapse = $synapse;
        $this->synapseInterface = $interface;
    }

    public function start() {

    }

    public function getSynapse(): Synapse {
        return $this->synapse;
    }

    public function emergencyShutdown() {
    }

    public function setName(string $name) {
    }

    public function process(): void {

    }

    public function close(Player $player, string $reason = 'unknown reason') {
    }

    public function putPacket(Player $player, DataPacket $packet, bool $needACK = false, bool $immediate = true) {
        if (!$player->isClosed()) {
            $pk = new RedirectPacket();
            $pk->uuid = $player->getUniqueId();
            $pk->direct = $immediate;
            if (!$packet->isEncoded) {
                $packet->encode();
            }
            $pk->mcpeBuffer = $packet->buffer;
            $this->synapseInterface->putPacket($pk);
        }
        return null;
    }

    public function shutdown() {
    }
}

File: src/main/java/org/itxtech/synapseapi/network/SynLibInterface.java
package org.itxtech.synapseapi.network;

import cn.nukkit.Player;
import cn.nukkit.Server;
import cn.nukkit.network.SourceInterface;
import cn.nukkit.network.protocol.BatchPacket;
import cn.nukkit.network.protocol.DataPacket;
import cn.nukkit.network.session.NetworkPlayerSession;
import cn.nukkit.utils.Binary;
import org.itxtech.synapseapi.network.protocol.spp.RedirectPacket;

import java.net.InetSocketAddress;

/**
 * Created by boybook on 16/6/24.
 */
public class SynLibInterface implements SourceInterface {

    private final SynapseInterface synapseInterface;

    public SynLibInterface(SynapseInterface synapseInterface) {
        this.synapseInterface = synapseInterface;
    }

    @Override
    public int getNetworkLatency(Player player) {
        return 0;
    }

    @Override
    public void emergencyShutdown() {
    }

    @Override
    public void setName(String name) {
    }

    @Override
    public Integer putPacket(Player player, DataPacket packet) {
        return this.putPacket(player, packet, false);
    }

    @Override
    public Integer putPacket(Player player, DataPacket packet, boolean needACK) {
        return this.putPacket(player, packet, needACK, false);
    }

    @Override
    public Integer putPacket(Player player, DataPacket packet, boolean needACK, boolean immediate) {
        RedirectPacket pk = new RedirectPacket();
        pk.uuid = player.getUniqueId();
        pk.direct = immediate;
        pk.mcpeBuffer = packet instanceof BatchPacket ? Binary.appendBytes((byte) 0xfe, ((BatchPacket) packet).payload) : packet.getBuffer();

        if (pk.mcpeBuffer.length >= 5242880) {
            Server.getInstance().getLogger().error("[Synapse] Too big packet! (pid: " + packet.pid() + ", player: " + player.getName() + ')');
        } else {
            this.synapseInterface.putPacket(pk);
        }
        return 0;
    }

    @Override
    public NetworkPlayerSession getSession(InetSocketAddress inetSocketAddress) {
        return null;
    }

    @Override
    public boolean process() {
        return false;
    }

    @Override
    public void close(Player player, String reason) {
    }

    @Override
    public void close(Player player) {
    }

    @Override
    public void shutdown() {
    }
}
